src/snowflake/connector/connection.py

```diff
@@ -1541,13 +1543,27 @@ def is_an_error(status: QueryStatus) -> bool:
 
     def _all_async_queries_finished(self) -> bool:
         """Checks whether all async queries started by this Connection have finished executing."""
-        queries = copy.copy(
-            self._async_sfqids
-        )  # get_query_status might update _async_sfqids, let's copy the list
-        finished_async_queries = (
-            not self.is_still_running(self.get_query_status(q)) for q in queries
-        )
-        return all(finished_async_queries)
+
+        queries = list(reversed(self._async_sfqids.keys()))
+        if not queries:
+            return True
+        num_workers = min(self.client_prefetch_threads, len(queries))
+        found_unfinished_query = False
+
+        def check_status_helper(
```

Comment:
If we want to avoid submitting all checks at the same time, we could also use a `callback` like:
```
with ThreadPoolExecutor(num_workers) as tpe:
    def check_status_callback(future: Future):
        nonlocal found_unfinished_query, tpe
        status = future.result()
        if self.is_still_running(status):
            found_unfinished_query = True
        if not found_unfinished_query and queries:
            try:
                sfq_id = queries.pop()
                tpe.submit(self.get_query_status, sfq_id).add_done_callback(check_status_callback)
            except IndexError:
                pass  # handle racing condition
```